### PR TITLE
removed matrix_cl<var>

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Removes `matrix_cl<var>` specialization. It will be replaced by `var_value<matrix_cl>`. Rev `cholesky_decompose` (the only place `matrix_cl<var>` was used) was modified to use two `matrix_cl<double>`s instead.

This PR also fixes a bug in kernel generator that caused generated kernel not to compile if the same expresion was present on both sides of an assignment. 

## Tests
Removes tests for removed functionalities. Added a test for the kernel generator bug. Re-added rev opencl `cholesky_decompose` test, which was deleted by mistake in #1424

## Side Effects
None.

## Release notes

Removed `matrix_cl<var>` specialization.

## Checklist

- [ ] Math issue #(issue number)

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/rev/fun/cholesky_decompose.hpp [hunk 1]
```diff
@@ -331,34 +336,30 @@ class cholesky_opencl : public vari {
...
-      R.adj() = R.adj() - transpose(C.adj()) * B.val() - D.adj() * R.val();
-      D.adj() = diagonal_multiply(D.adj(), 0.5);
+      R_adj = R_adj - transpose(C_adj) * B_val - D_adj * R_val;
+      D_adj = diagonal_multiply(D_adj, 0.5);
 
-      L.adj().sub_block(R.adj(), 0, 0, j, 0, k_j_ind, j);
-      L.adj().sub_block(D.adj(), 0, 0, j, j, k_j_ind, k_j_ind);
-      L.adj().sub_block(B.adj(), 0, 0, k, 0, m_k_ind, j);
-      L.adj().sub_block(C.adj(), 0, 0, k, j, m_k_ind, k_j_ind);
+      block(L_adj, j, j, k_j_ind, k_j_ind) = D_adj;
```
Comment (rok-cesnovar) [inline stan/math/rev/fun/cholesky_decompose.hpp:357]: The rest are now handled by the kernel generator right?

Comment (t4c1) [reply stan/math/rev/fun/cholesky_decompose.hpp:357]: The rest use  only kernel gneerator functionis, so tzhey can use the blocks in place. This one uses some other functions that only work on matrix_cl, so we still need copys.

File: stan/math/rev/fun/cholesky_decompose.hpp [hunk 2]
```diff
@@ -331,34 +335,30 @@ class cholesky_opencl : public vari {
...
     }
-    L.view(matrix_cl_view::Lower);
-    L_adj_cpu = packed_copy(L);
+    L_adj.view(matrix_cl_view::Lower);
+    std::vector<double> L_adj_cpu2 = packed_copy(L_adj);
     for (size_type j = 0; j < packed_size; ++j) {
-      vari_ref_A_[j]->adj_ += L_adj_cpu[j];
+      vari_ref_A_[j]->adj_ += L_adj_cpu2[j];
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/rev/fun/cholesky_decompose.hpp]: done

**Review by rok-cesnovar — APPROVED** [review]
Look good. Thanks!

**Pull request merged**